File: src/sample07/br/edu/ifpi/poo/banco/cadastros/Banco.java
package sample07.br.edu.ifpi.poo.banco.cadastros;

import sample07.br.edu.ifpi.poo.banco.entidade.AplicacaoException;
import sample07.br.edu.ifpi.poo.banco.entidade.Conta;
import sample07.br.edu.ifpi.poo.banco.entidade.Poupanca;

public class Banco {
	private Conta[] contas;
	private int indice = 0;

	public Banco() {
		contas = new Conta[50];
	}

	public Banco(int tamanho) {
		contas = new Conta[tamanho];
	}

	public void inserir(Conta c) {
		if (indice >= contas.length) {
			throw new AplicacaoException("Número máximo de contas atingido");
		}
		contas[indice] = c;
		indice++;
	}

	public Conta consultar(String numero) throws AplicacaoException {
		Conta c = null;
		for (int i = 0; i < indice; i++) {
			if (contas[i].getNumero().equals(numero)) {
				c = contas[i];
				break;
			}
		}

		if (c == null) throw new AplicacaoException("Nenhuma conta foi encontrada com o número " + numero);
		return c;
	}

	private int consultarIndice(String numero) {
		int pos = -1;
		for (int i = 0; i < indice; i++) {
			if (contas[i].getNumero().equals(numero)) {
				pos = i;
				break;
			}
		}
		return pos;
	}

	public void debitar(String numero, double valor) throws AplicacaoException {
		Conta c;
		c = consultar(numero);
		if (c == null) {
			throw new AplicacaoException("Conta inexistente.");
		}
		
		c.sacar(valor);
	}

	public void creditar(String numero, double valor) throws AplicacaoException {
		Conta c;
		c = consultar(numero);
		if (c == null) throw new AplicacaoException("Conta não encontrada");

		c.depositar(valor);
	}
	
	public void transferir(String numCredito, String numDebito, double valor) {
		Conta contaCredito = consultar(numCredito);
		Conta contaDebito = consultar(numDebito);

		if (contaCredito == null || contaDebito == null) {
			throw new AplicacaoException("As duas contas devem existir.");
		}
		contaCredito.transferir(contaDebito, valor);
	}

	public void alterar(Conta c) {
		int indice = consultarIndice(c.getNumero());
		if (indice != -1) {
			contas[indice] = c;
		}

	}

	public void excluir(String numero) {
		int posicao = consultarIndice(numero);
		if (posicao == -1) {
			throw new AplicacaoException("Índice não encontrado");
		}

		for (int i = posicao; i < indice; i++) {
			contas[i] = contas[i + 1];
		}

		indice--;
	}
	
	public void renderJuros(String numero) {
		Conta c = consultar(numero);
		if (!(c instanceof Poupanca)) {
			throw new AplicacaoException("Poupançaa não encontrada");
		}
		((Poupanca) c).renderJuros();
	}
	
	public int retornaQuantidade() {
		return indice;
	}
	
	public double retornaValorTotal() {
		double total = 0;
		for (int i = 0; i < indice; i++) {
			total += contas[i].getSaldo();
		}
		
		return total;
	}
	
	public double retornaMediaValores() {
		return retornaValorTotal()/retornaQuantidade();
	}
}

File: src/sample09/Banco.ts
import { Conta } from "./conta/Conta";
import { Poupanca } from "./conta/Poupanca";
import { ContaInexistenteError, PoupancaInvalidaError } from "./conta/Errors";

export class Banco {
  private _contas: Conta[] = [];

  inserir(conta: Conta): void {
    this.consultar(conta.numero);
    this._contas.push(conta);
  }

  consultar(numero: String): Conta {
    let contaConsultada: Conta | null = null;
    for (let conta of this._contas) {
      if (conta.numero == numero) {
        contaConsultada = conta;
        break;
      }
    }

    if (!contaConsultada) {
      throw new ContaInexistenteError();
    }
    return contaConsultada;
  }

  private consultarPorIndice(numero: String): number {
    let indice: number = -1;
    for (let i = 0; i < this._contas.length; i++) {
      if (this._contas[i].numero == numero) {
        indice = i;
        break;
      }
    }
    return indice;
  }

  alterar(conta: Conta): void {
    let indice = this.consultarPorIndice(conta.numero);
    this._contas[indice] = conta;
  }

  excluir(numero: string): void {
    let indice: number = this.consultarPorIndice(numero);

    if (indice != -1) {
      for (let i: number = indice; i < this._contas.length; i++) {
        this._contas[i] = this._contas[i + 1];
      }

      this._contas.pop();
    }
  }

  depositar(numero: String, valor: number): void {
    let contaConsultada = this.consultar(numero);
    contaConsultada.depositar(valor);
  }

  sacar(numero: String, valor: number): void {
    let contaConsultada = this.consultar(numero);
    contaConsultada.sacar(valor);
  }

  transferir(numeroDebito: string, numeroCredito: string, valor: number) {
    let contaCredito: Conta = this.consultar(numeroCredito);
    let contaDebito: Conta = this.consultar(numeroDebito);
    contaDebito.transferir(contaCredito, valor);
  }

  calcularQuantidadeContas(): number {
    return this._contas.length;
  }

  calcularTotalSaldos(): number {
    let totalSaldo: number = 0;
    for (let conta of this._contas) {
      totalSaldo += conta.saldo;
    }

    return totalSaldo;
  }

  calcularMediaSaldos() {
    return this.calcularTotalSaldos() / this.calcularQuantidadeContas();
  }

  renderJuros(numero: String) {
    let contaConsultada: Conta = this.consultar(numero);
    console.log(contaConsultada instanceof Poupanca)

    if (!(contaConsultada instanceof Poupanca)) {
      throw new PoupancaInvalidaError();
    }

    if (contaConsultada instanceof Poupanca) {
      let poupanca: Poupanca = <Poupanca>contaConsultada;
      poupanca.renderJuros();
    }
  }
}
